keras/utils/generic_utils.py

```diff
@@ -286,23 +287,34 @@ class Progbar(object):
 
     # Arguments
         target: Total number of steps expected, None if unknown.
+        width: Progress bar width on screen.
+        verbose: Verbosity mode, 0 (silent), 1 (verbose), 2 (semi-verbose)
+        stateful_metrics: Iterable of string names of metrics that
+            should *not* be averaged over time. Metrics in this list
+            will be displayed as-is. All others will be averaged
+            by the progbar before display.
         interval: Minimum visual progress update interval (in seconds).
     """
 
-    def __init__(self, target, width=30, verbose=1, interval=0.05):
-        self.width = width
+    def __init__(self, target, width=30, verbose=1, interval=0.05,
+                 stateful_metrics=None):
```

Comment:
I think this change should be the other way around. `stateful_metrics` is weirdly specific for a progbar... We should do away with all the explicit progbar calls and instead make a callback call. Code for something like this is available at https://github.com/bstriner/keras-tqdm. On this topic, have you seen [tqdm](https://github.com/tqdm/tqdm)? 

```
class tqdm(object):
  """
  Decorate an iterable object, returning an iterator which acts exactly
  like the original iterable, but prints a dynamically updating
  progressbar every time a value is requested.

A much better way to handle this is the way [tqdm](https://github.com/tqdm/tqdm) does it, or using tqdm explicitly:
  """
  def __init__(self, iterable=None, desc=None, total=None, leave=True,
               file=None, ncols=None, mininterval=0.1,
               maxinterval=10.0, miniters=None, ascii=None, disable=False,
               unit='it', unit_scale=False, dynamic_ncols=False,
               smoothing=0.3, bar_format=None, initial=0, position=None,
               postfix=None):
```

It works much more nicely than this progbar is compatible with ipython notebooks, and is completely optional when imported as follows:

```python

# progress bars https://github.com/tqdm/tqdm
# import tqdm without enforcing it as a dependency
try:
    from tqdm import tqdm
except ImportError:

    def tqdm(*args, **kwargs):
        if args:
            return args[0]
        return kwargs.get('iterable', None)
```

Reply:
I think this is out of scope, the goal of the PR is to support stateful metrics.

A paradigm shift in the PR (if that's the route) should be a follow on PR.

The ProgBar change leads to no change in UX, but the stateful metrics is a big improvement.

Reply:
@briannemsick Sorry, I wasn't clear about my main point! 

1. Rather than actually including tqdm (which is just a convenient option), the point is the superior API UX style of tqdm for this stateful task. The API is worth using for inspiration when considering the stateful metrics, which must be updated at every batch.
2. It should be a callback call and not a special init variable that gets passed around

Reply:
The current progbar is fine. We're not changing the existing API, we just extend it in a really simple way to support a new use case.

Reply:
> The current progbar is fine. We're not changing the existing API, we just extend it in a really simple way to support a new use case.

Fair enough, but rather than a different progbar, please allow me to suggest a different stateful metric design. In other words, consider a stateful metric which is a decorated iterator implementing `next()`, `update()`, `clear()`, etc.

Reply:
Thanks for the issue links, I hadn't seen them!

1. Can [tf.metrics](https://www.tensorflow.org/api_docs/python/tf/metrics) and [tf.contrib.metrics](https://www.tensorflow.org/api_docs/python/tf/contrib/metrics) mentioned in the linked discussion easily be adapted to the proposed API?
2. Can we design the API UX to look like [boltons.statsutils](http://boltons.readthedocs.io/en/latest/statsutils.html)?

Item 2 is what I was really think of and it is a much clearer example than tqdm.